File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/noughtsandcrosses/domain/NoughtsAndCrossesGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.ArrayList;
import java.util.List;

import fj.F2;

import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;

public class NoughtsAndCrossesGame extends CellsGame<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove, NoughtsAndCrossesGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };
    public static Position offset2[] = {
            new Position(0, 0),
            new Position(1, 1),
            new Position(1, 1),
            new Position(0, 0),
    };
    public static int dirs[] = {1, 0, 3, 2};

    public char[] objArray;
    public char get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public char get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, char obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, char obj) {
        set(p.row, p.col, obj);
    }

    public List<Position> noughts = new ArrayList<>();
    public char chMax;

    public NoughtsAndCrossesGame(List<String> layout, char chMax, GameInterface<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove, NoughtsAndCrossesGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        this.chMax = chMax;
        size = new Position(layout.size(), layout.get(0).length());
        objArray = new char[rows() * cols()];
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = str.charAt(c);
                if (ch == 'O') noughts.add(p);
                set(p, ch == 'O' ? ' ' : ch);
            }
        }

        NoughtsAndCrossesGameState state = new NoughtsAndCrossesGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(NoughtsAndCrossesGameMove move, F2<NoughtsAndCrossesGameState, NoughtsAndCrossesGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        NoughtsAndCrossesGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(NoughtsAndCrossesGameMove move) {
        return changeObject(move, NoughtsAndCrossesGameState::switchObject);
    }

    public boolean setObject(NoughtsAndCrossesGameMove move) {
        return changeObject(move, NoughtsAndCrossesGameState::setObject);
    }

    public char getObject(Position p) {
        return state().get(p);
    }

    public char getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState getRowState(int row) {
        return state().row2state[row];
    }

    public HintState getColState(int col) {
        return state().col2state[col];
    }

    public HintState getPosState(Position p) {
        return defaultIfNull(state().pos2state.get(p), HintState.Normal);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/noughtsandcrosses/NoughtsAndCrossesGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.HintState
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class NoughtsAndCrossesGame(layout: List<String>, var chMax: Char, gi: GameInterface<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove, NoughtsAndCrossesGameState>, gdi: GameDocumentInterface) : CellsGame<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove, NoughtsAndCrossesGameState>(gi, gdi) {
    companion object {
        val offset = Position.Directions4
        val offset2 = arrayOf(
            Position(0, 0),
            Position(1, 1),
            Position(1, 1),
            Position(0, 0)
        )
        val dirs = intArrayOf(1, 0, 3, 2)
    }

    val objArray: CharArray

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: Char) {objArray[row * cols + col] = obj}
    operator fun set(p: Position, obj: Char) {this[p.row, p.col] = obj}

    val noughts = mutableListOf<Position>()

    init {
        size = Position(layout.size, layout[0].length)
        objArray = CharArray(rows * cols)
        for (r in 0 until rows) {
            var str = layout[r]
            for (c in 0 until cols) {
                val p = Position(r, c)
                val ch = str[c]
                if (ch == 'O') noughts.add(p)
                this[p] = if (ch == 'O') ' ' else ch
            }
        }
        val state = NoughtsAndCrossesGameState(this)
        levelInitialized(state)
    }

    fun getObject(p: Position) = currentState[p]
    fun getObject(row: Int, col: Int) = currentState[row, col]
    fun row2state(row: Int) = currentState.row2state[row]
    fun col2state(col: Int) = currentState.col2state[col]
    fun pos2state(p: Position) = currentState.pos2state[p] ?: HintState.Normal
}
